astropy/time/core.py

```diff
@@ -276,11 +266,57 @@ def new_like(self, cols, length, metadata_conflicts='warn', name=None):
         return out
 
 
-class TimeDeltaInfo(TimeInfo):
-    _represent_as_dict_extra_attrs = ('format', 'scale')
+class TimeInfo(TimeInfoBase):
+    def _represent_as_dict(self, attrs=None):
+        """Get the values for the parent ``attrs`` and return as a dict.
+
+        By default, uses '_represent_as_dict_attrs'.
+        """
+        map = super()._represent_as_dict(attrs=attrs)
+
+        # The datetime64 format requires special handling for ECSV (see #12840).
+        # The `value` has numpy dtype datetime64 but this is not an allowed
+        # datatype for ECSV. Instead convert to a string representation.
+        if (self._serialize_context == 'ecsv'
+                and map['format'] == 'datetime64'
+                and 'value' in map):
+            map['value'] = map['value'].astype('U')
+
+        # The datetime format is serialized as ISO with no loss of precision.
+        if map['format'] == 'datetime':
+            map['value'] = np.array([val.isoformat() for val in map['value']])
+
+        return map
 
     def _construct_from_dict(self, map):
-        return self._construct_from_dict_base(map)
+        # See comment above. May need to convert string back to datetime64.
+        # Note that _serialize_context is not set here so we just look for the
+        # string value directly.
+        if (map['format'] == 'datetime64'
+                and 'value' in map
+                and map['value'].dtype.kind == 'U'):
+            map['value'] = map['value'].astype('datetime64')
+
+        # Convert back to datetime objects for datetime format.
+        if map['format'] == 'datetime':
+            from datetime import datetime
+            map['value'] = [datetime.fromisoformat(val) for val in map['value']]
```

Comment:
Similarly, this would need to be `np.vectorize(datetime.fromisoformat)(map['value'])`